usr/lib/common/lock_sess_mgr.c

```diff
@@ -614,22 +616,29 @@ CK_RV session_mgr_get_op_state(SESSION *sess,
             return CKR_STATE_UNSAVEABLE;
         }
         active_ops++;
-        if (op_data != NULL) {
-            TRACE_ERROR("%s\n", ock_err(ERR_STATE_UNSAVEABLE));
-            return CKR_STATE_UNSAVEABLE;
-        }
         op_data_len = sizeof(OP_STATE_DATA) +
             sizeof(ENCR_DECR_CONTEXT) +
             sess->encr_ctx.context_len + sess->encr_ctx.mech.ulParameterLen;
+        all_data_len += op_data_len;
 
         if (length_only == FALSE) {
             op_data = (OP_STATE_DATA *) data;
 
-            if (*data_len < op_data_len) {
+            if (max_data_len < op_data_len) {
                 TRACE_ERROR("%s\n", ock_err(ERR_BUFFER_TOO_SMALL));
                  return CKR_BUFFER_TOO_SMALL;
             }
 
+            memset(op_data, 0, sizeof(*op_data));
+#ifdef PACKAGE_VERSION
+            strncpy((char *)op_data->library_version, PACKAGE_VERSION,
+                    sizeof(op_data->library_version));
+#endif
+            memcpy(op_data->manufacturerID,
+                   tokdata->nv_token_data->token_info.manufacturerID,
+                   sizeof(op_data->manufacturerID));
+            memcpy(op_data->model, tokdata->nv_token_data->token_info.model,
+                   sizeof(op_data->model));
```

Comment:
These lines are copied 3 times. Wouldn't it make sense to make a function for it?

Reply:
Yes actually the whole code around these lines is pretty much the same all 5 times... It just differs slightly by the context type it operates on.... So maybe the code should be refactored to have that lines only once, either in a macro or in a function.

I put this on my TODO list for a later PR.....